MoreLinq/Experimental/Await.cs

```diff
@@ -531,149 +610,82 @@ IEnumerable<TResult> _(int maxConcurrency, TaskScheduler scheduler, bool ordered
-                cancellationToken.Register(() => cancellationTaskSource.TrySetResult(true));
-
-                var tasks = new List<(T Item, Task<TResult> Task)>();
-
-                for (var i = 0; i < maxConcurrency; i++)
+                void OnPendingCompleted()
                 {
-                    if (!reader.TryRead(out var item))
-                        break;
-                    tasks.Add((item, taskSelector(item)));
+                    if (Interlocked.Decrement(ref pendingCount) == 0)
+                        onEnd();
                 }
 
-                while (tasks.Count > 0)
+                var semaphore = maxConcurrency is int count
+                              ? new SemaphoreSlim(count, count)
+                              : null;
+
+                while (enumerator.MoveNext())
                 {
-                    // Task.WaitAny is synchronous and blocking but allows the
-                    // waiting to be cancelled via a CancellationToken.
-                    // Task.WhenAny can be awaited so it is better since the
-                    // thread won't be blocked and can return to the pool.
-                    // However, it doesn't support cancellation so instead a
-                    // task is built on top of the CancellationToken that
-                    // completes when the CancellationToken trips.
-                    //
-                    // Also, Task.WhenAny returns the task (Task) object that
-                    // completed but task objects may not be unique due to
-                    // caching, e.g.:
-                    //
-                    //     async Task<bool> Foo() => true;
-                    //     async Task<bool> Bar() => true;
-                    //     var foo = Foo();
-                    //     var bar = Bar();
-                    //     var same = foo.Equals(bar); // == true
-                    //
-                    // In this case, the task returned by Task.WhenAny will
-                    // match `foo` and `bar`:
-                    //
-                    //     var done = Task.WhenAny(foo, bar);
-                    //
-                    // Logically speaking, the uniqueness of a task does not
-                    // matter but here it does, especially when Await (the main
-                    // user of CollectAsync) needs to return results ordered.
-                    // Fortunately, we compose our own task on top of the
-                    // original that links each item with the task result and as
-                    // a consequence generate new and unique task objects.
-
-                    var completedTask = await
-                        Task.WhenAny(tasks.Select(it => (Task) it.Task).Concat(cancellationTaskSource.Task))
-                            .ConfigureAwait(continueOnCapturedContext: false);
-
-                    if (completedTask == cancellationTaskSource.Task)
+                    if (semaphore != null)
```

Comment:
I think a wrapper class would make the code simpler:

```csharp
class SemaphoreSlimWrapper {
  private readonly SemaphoreSlim semaphore;
  public SemaphoreSlimWrapper(int? maxConcurrency) {
      semaphore = maxConcurrency is int count
                              ? new SemaphoreSlim(count, count)
                             : null;
  }
  public Task WaitAsync() => semaphore?.WaitAsync() ?? Task.CompletedTask;
  public void Release() => semaphore?.Release();
}
```

And then this if can be avoided.

Reply:
Let me know what you make of it now with e956b461d198a601d0cc891f4a821bcfc9adb03f.